[ML] Hyperparameter importance

valeriy42 opened the pull request:
This PR implements hyperparameter importance calculation using ANOVA decomposition of the Gaussian process model used in the Bayesian optimization routine. The results is recorded in the model metadata structure. For convenience, I record the anova variance as an absolute value as well as a fraction of the total variation. This "relative importance" simplifies the comparison between different parameters. 

Additionally, I record the best value for the individual hyperparameter s.t. it can be used in a follow-up job configuration and doesn't have to be looked up separately.

Example output:
```json
"hyperparameter_importance": [
        {
            "name": "downsample_factor",
            "value": 0.35355339059327386,
            "absolute_importance": 0.123,
            "relative_importance": 0.9123
        },
        {
            "name": "alpha",
            "value": 0.010766199582740555,
            "absolute_importance": 0.0123,
            "relative_importance": 0.0321
        }
    ]
```

In order to verify the analytical formulas, I implemented Sobol sequence random number generation, which leads to lower variance for Monte-Carlo estimations.

Closes #1594

tveasey commented on lib/maths/CBayesianOptimisation.cc:
```diff
@@ -189,6 +205,166 @@ CBayesianOptimisation::maximumExpectedImprovement() {
+            sum += prodIj(i, j);
+        }
+    }
+    double variance{std::pow(m_KernelParameters(0), 4) * sum -
+                    CTools::pow2(this->anovaConstantFactor(Kinvf))};
```
This will have rather bad cancellation errors if `m_KernelParameters(0)` is large and we know that the constant factor is proportional to `m_KernelParameters(0)^2`. Let's discuss numerics a bit offline.

tveasey replied:
Also, I think we should be bounding variances to >= 0.0.

valeriy42 replied:
I bounded the variance.

tveasey commented on lib/maths/CBayesianOptimisation.cc:
```diff
@@ -58,6 +59,21 @@ double stableNormPdf(double z) {
 // narrow deep valley that the Gaussian Process hasn't sampled.
 const double MINIMUM_KERNEL_SCALE_FOR_EXPECTATION_MAXIMISATION{1e-8};
+
+double stableProdd(double theta1, double x) {
```
Could we maybe think of a better name for these two functions, i.e. I think that these should reflect the fact they are coming from integrals of the kernel.

valeriy42 replied:
Alright, I call them `integrate1dKernel` and `integrate1dKernelProduct`. Do you think it's better?

tveasey replied:
Sounds good.

tveasey reviewed (commented):
I've done a first pass through. Overall looks great. I've focused mainly on some robustness suggestions. The tests passing are pretty convincing that the implementation is right, but I think it might be worth seeing if we can tighten up the numerics in a couple of places. Let's talk about this a bit offline.

valeriy42 commented on lib/maths/CBoostedTreeFactory.cc:
```diff
@@ -180,49 +185,60 @@ void CBoostedTreeFactory::initializeHyperparameterOptimisation() const {
     // less than p_1, this translates to using log parameter values.
 
     CBayesianOptimisation::TDoubleDoublePrVec boundingBox;
+    THyperparametersVec hyperparameters;
```
Good idea, fixed.

valeriy42 commented on lib/maths/CBoostedTreeImpl.cc:
```diff
@@ -1233,39 +1233,70 @@ bool CBoostedTreeImpl::selectNextHyperparameters(const TMeanVarAccumulator& loss
-    if (m_FeatureBagFractionOverride == boost::none) {
-        parameters(i++) = m_FeatureBagFraction;
+    for (int i = 0; i < static_cast<int>(m_TunableHyperparameters.size()); ++i) {
+        switch (m_TunableHyperparameters[static_cast<std::size_t>(i)]) {
```
I am getting sign-conversion warnings without the static cast, that's why I added it even if it looks ugly.

droberts195 replied:
Could you just make `i` a `std::size_t` in its declaration in the `for` loop instead?  That's what we'd usually do unless there's a really pressing need to have a signed loop variable.

tveasey replied:
> Could you just make i a std::size_t

I'd second this, but it be to suppress warnings Valeriy is seeing from `parameters` which is an Eigen vector and which expects integer indexes. I never see warnings on function argument type coercion. Also, I don't recall seeing them from our linux build. I wonder which compiler version this is?

valeriy42 replied:
I am getting the `clang` wornings in the VS Code IDE. I'll change and check the gcc compiler workings.

valeriy42 commented on lib/maths/CSampling.cc:
```diff
@@ -820,6 +823,26 @@ void CSampling::gammaSampleQuantiles(double shape, double rate, std::size_t n, T
+    quasi_random_gen_t gen(engine, boost::uniform_01<double>());
+    {
+        samples.resize(n, TDoubleVec(dim));
+        CDenseVector<double> sample(dim);
```
missed this one. Thank you.

valeriy42 commented on lib/maths/CSampling.cc:
```diff
@@ -820,6 +823,26 @@ void CSampling::gammaSampleQuantiles(double shape, double rate, std::size_t n, T
+        return;
+    }
+    typedef boost::variate_generator<boost::random::sobol&, boost::uniform_01<double>> quasi_random_gen_t;
+    boost::random::sobol engine(dim);
```
fixed

valeriy42 commented on lib/maths/CBayesianOptimisation.cc:
```diff
@@ -189,6 +208,166 @@ CBayesianOptimisation::maximumExpectedImprovement() {
+    TVector scaledKernelParameters{this->scaledKernelParameters()};
+    auto prodXt = [this, &scaledKernelParameters](const TVector& x, int t) {
+        double prod{1.0};
+        for (int d = 0; d < this->m_MinBoundary.size(); ++d) {
```
Fixed

valeriy42 commented on lib/maths/CBayesianOptimisation.cc:
```diff
@@ -189,6 +208,166 @@ CBayesianOptimisation::maximumExpectedImprovement() {
+    return Kxn.transpose() * Kinvf;
+}
+
+double CBayesianOptimisation::evaluate1D(const TVector& Kinvf, double input, int dimension) const {
```
Fixed.

valeriy42 commented on lib/maths/CBoostedTreeImpl.cc:
```diff
@@ -1233,39 +1233,70 @@ bool CBoostedTreeImpl::selectNextHyperparameters(const TMeanVarAccumulator& loss
+            parameters(i) = m_Regularization.softTreeDepthTolerance();
+            break;
+        }
     }
```
Good point! I added the term in `estimateMemoryUsage` and implemented `initializeTunableHyperparameters` to (re-) populate the list after restoring the object. This way we don't have to change the schema. Please, let me know if I have forgot something.

tveasey commented on lib/maths/CBayesianOptimisation.cc:
```diff
@@ -189,6 +205,176 @@ CBayesianOptimisation::maximumExpectedImprovement() {
+        }
+        return prod;
+    };
+    double f0{this->anovaConstantFactor(Kinvf)};
```
As before, since we're differencing numbers proportional to theta_0^4 in some places I'm a little worried about cancellation errors. Using here as an example of what I have in mind (same approach applies elsewhere), I would propose introducing a scale. For example, the following changes to:
```
    double theta02{CTools::pow2(m_KernelParameters(0))};
    double scale{std::max(1.0, theta02)};
    double f0{this->anovaConstantFactor(Kinvf) / scale};
    double sum{0.0};
    input = (input - m_MinBoundary(dimension)) /
            (m_MaxBoundary(dimension) - m_MinBoundary(dimension));
    for (std::size_t i = 0; i < m_FunctionMeanValues.size(); ++i) {
        TVector x{this->transformTo01(m_FunctionMeanValues[i].first)};
        sum += Kinvf(static_cast<int>(i)) *
               std::exp(-(CTools::pow2(scaledKernelParameters[dimension]) +
                          MINIMUM_KERNEL_COORDINATE_DISTANCE_SCALE) *
                        CTools::pow2(input - x(dimension))) *
               prodXt(x, dimension);
    }
    return scale * (theta02 / scale * sum - f0);
```
I feel this doesn't complicate matters too much and deals with the edge case. Of course it only really matters if `|f0| >> |CTools::pow2(m_KernelParameters(0)) * sum - f0|`. Also these are approximations of the quantities we care about anyway and given we're using double precision we probably get away with it. I'll leave it to you to decide whether to roll out these sorts of change.

valeriy42 replied:
Better safe than sorry. I added your suggestion to the relevant functions.

tveasey reviewed (commented):
I did another pass. This is looking pretty close. The remaining things are mainly cosmetic. I also made a concrete suggestion on dealing with cancellation error.

tveasey commented:
One last thing: failures of
```
lib.maths.CBoostedTreeTest.CBoostedTreeTest.testCategoricalRegressors
lib.maths.CBoostedTreeTest.CBoostedTreeTest.testBinomialLogisticRegression
lib.api.CDataFrameAnalyzerTrainingTest.CDataFrameAnalyzerTrainingTest.testRunBoostedTreeRegressionTrainingWithStateRecovery
```
indicate some error since these changes shouldn't affect any existing tests. The most likely cause is the refactor of hyperparameter handling. I didn't spot any obvious issue, but this needs resolving before merge.

valeriy42 commented on include/maths/CBayesianOptimisation.h:
```diff
@@ -93,6 +93,31 @@ class MATHS_EXPORT CBayesianOptimisation {
+    double evaluate(const TVector& input) const;
+
+    //! Compute the marginalized value of the Gaussian process in the dimension
+    //! \p t for the values \p input.
```
Fixed

valeriy42 commented on include/maths/CBoostedTreeUtils.h:
```diff
@@ -31,6 +31,33 @@ using TAlignedMemoryMappedFloatVector =
+    E_Gamma,
+    E_Lambda,
+    E_SoftTreeDepthLimit,
+    E_SoftTreeDepthTolerance,
+    E_LastHyperparameter //put all new hyperparameter before this one
+};
```
Fixed

valeriy42 commented on include/maths/CBoostedTreeUtils.h:
```diff
@@ -31,6 +31,33 @@ using TAlignedMemoryMappedFloatVector =
+                              double relativeImportance)
+        : s_Hyperparameter(hyperparameter), s_Value(value),
+          s_AbsoluteImportance(absoluteImportance),
+          s_RelativeImportance(relativeImportance){};
```
fixed

valeriy42 commented on lib/maths/CBoostedTreeFactory.cc:
```diff
@@ -180,49 +183,81 @@ void CBoostedTreeFactory::initializeHyperparameterOptimisation() const {
+                boundingBox.emplace_back(MIN_FEATURE_BAG_FRACTION, MAX_FEATURE_BAG_FRACTION);
+            }
+            break;
+        case E_LastHyperparameter:
+            // only a placeholder
+            break;
```
Fixed

valeriy42 commented on lib/maths/CBoostedTreeImpl.cc:
```diff
@@ -1283,36 +1293,50 @@ bool CBoostedTreeImpl::selectNextHyperparameters(const TMeanVarAccumulator& loss
+        case E_SoftTreeDepthTolerance:
+            m_Regularization.softTreeDepthTolerance(parameters(i));
+            break;
+        case E_LastHyperparameter:
+            break;
+        }
```
removed

valeriy42 commented on lib/maths/CBoostedTreeImpl.cc:
```diff
@@ -1233,40 +1236,47 @@ bool CBoostedTreeImpl::selectNextHyperparameters(const TMeanVarAccumulator& loss
+        case E_SoftTreeDepthTolerance:
+            parameters(i) = m_Regularization.softTreeDepthTolerance();
+            break;
+        case E_LastHyperparameter:
+            break;
+        }
```
removed

valeriy42 commented on lib/maths/CBoostedTreeImpl.cc:
```diff
@@ -1390,6 +1414,62 @@ void CBoostedTreeImpl::recordHyperparameters() {
+                m_TunableHyperparameters.emplace_back(E_FeatureBagFraction);
+            }
+            break;
+        case E_LastHyperparameter:
+            // only a placeholder
+            break;
```
removed

valeriy42 commented on lib/maths/CSampling.cc:
```diff
@@ -820,6 +823,29 @@ void CSampling::gammaSampleQuantiles(double shape, double rate, std::size_t n, T
+    }
+
+    try {
+        typedef boost::variate_generator<boost::random::sobol&, boost::uniform_01<double>> quasi_random_gen_t;
```
Missed this one. Thanks!

valeriy42 commented on lib/maths/CBoostedTreeFactory.cc:
```diff
@@ -82,6 +83,8 @@ std::size_t computeMaximumNumberTrees(double eta) {
 bool intervalIsEmpty(const CBoostedTreeFactory::TVector& interval) {
     return interval(MAX_REGULARIZER_INDEX) - interval(MIN_REGULARIZER_INDEX) == 0.0;
 }
+
+using THyperparametersVec = std::vector<boosted_tree_detail::EHyperparameters>;
```
Removed

valeriy42 commented:
Thank you for the review @tveasey . I addressed all your comments. It would be great if you could take another look. 

In the mean time I'll look at how to fix the unit test failures.

tveasey reviewed (approved):
Thanks for iterating on this. Good stuff! This LGTM once it passes CI.

valeriy42 commented:
retest
